Common++/header/IpAddresses.h

```diff
@@ -0,0 +1,377 @@
+	{
+		uint32_t addr;
+		memcpy(&addr, m_Bytes, sizeof(m_Bytes));
+		return addr;
+	}
+
+
+
+
+
+	/**
+	 * @class IPv6Address
+	 * Represents an IPv6 address (of type xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx).
+	 */
+	class IPv6Address
+	{
+	public:
+		/**
+		 * A default constructor that creates an instance of the class with unspecified/zero address
+		 */
+		IPv6Address() { memset(m_Bytes, 0, sizeof(m_Bytes)); }
+
+		/**
+		 * A constructor that creates an instance of the class out of 16-byte array.
+		 * @param[in] bytes The address as 16-byte array in network byte order
+		 */
+		IPv6Address(const uint8_t bytes[16]) { memcpy(m_Bytes, bytes, sizeof(m_Bytes)); }
+
+		/**
+		 * A constructor that creates an instance of the class out of std::string value
+		 * If the string doesn't represent a valid IPv6 address, an instance will store an unspecified address
+		 * @param[in] addressAsString The std::string representation of the address
+		 */
+		IPv6Address(const std::string& addrAsString);
+
+		/**
+		 * Returns a pointer to 16-byte array representing the IPv6 address
+		 */
+		const uint8_t* toBytes() const { return m_Bytes; }
+
+		/**
+		 * Returns a std::string representation of the address
+		 * @return A string representation of the address
+		 */
+		std::string toString() const;
+
+		/**
+		 * Determine whether the address is unspecified
+		 */
+		bool isUnspecified() const { return *this == Zero; }
+
+		/**
+		 * Overload of the equal-to operator
+		 */
+		bool operator==(const IPv6Address& rhs) const { return memcmp(toBytes(), rhs.toBytes(), sizeof(m_Bytes)) == 0; }
+
+		/**
+		 * Overload of the not-equal-to operator
+		 */
+		bool operator!=(const IPv6Address &rhs) const { return !(*this == rhs); }
+
+
+		// Following methods are placed for backward compatibility and will be deleted in the future
+
+		/**
+		 * Allocates a byte array and copies address value into it. Array deallocation is user responsibility
+		 * Notice this method is deprecated and will be deleted in the future
+		 * @param[in] arr A pointer to where array will be allocated
+		 * @param[out] length Returns the length in bytes of the array that was allocated
+		 */
+		#if __cplusplus > 201402L || _MSC_VER >= 1900
+		[[deprecated("This method is now unnecessary and added for backward compatibility. It will be deleted in the future")]]
+		#endif
+		void copyTo(uint8_t** arr, size_t& length) const;
+
+		/**
+		 * Gets a pointer to an already allocated byte array and copies the address value to it.
+		 * Notice this method is deprecated and will be deleted in the future
+		 * This method assumes array allocated size is at least 16 (the size of an IPv6 address)
+		 * @param[in] arr A pointer to the array which address will be copied to
+		 */
+		#if __cplusplus > 201402L || _MSC_VER >= 1900
+		[[deprecated("This method is now unnecessary and added for backward compatibility. It will be deleted in the future")]]
+		#endif
+		void copyTo(uint8_t* arr) const;
```

Comment:
Why are these methods unnecessary?

Reply:
`void copyTo(uint8_t** arr, size_t& length) const` allocates the memory internally. The user can use the different ways to allocate the memory such as `make_unique`, `make_shared` or something else. So let the user select the appropriate method.

`void copyTo(uint8_t* arr)` just replaces the `memcpy`, `std::copy`

Reply:
Regarding `void copyTo(uint8_t* arr)` - I know it's just a replacement for `memcpy` or `std::copy` but it gives a nice and easy-to-use interface for the user.

Regarding `void copyTo(uint8_t** arr, size_t& length) const` - if the user wants to allocate the memory himself he can use the other `copyTo()` method. This method is for users who wants to provide an empty pointer and let `IPv6Address` allocate the relevant amount of memory and do the copy.

I do agree that these 2 methods are not a must but they're giving a more easy-to-use interface for the users which is what PcapPlusPlus is basically all about 😃

Reply:
Ok, I will make the changes

Reply:
Done